jormungandr/src/network/p2p/topology.rs

```diff
@@ -2,171 +2,207 @@
+    // forgotten by the underlying poldercast implementation.
+    pub async fn list_quarantined(&self) -> Vec<ProfileInfo> {
+        self.lock.read().await.quarantine.quarantined_nodes()
     }
 
-    pub async fn list_available(&self) -> Vec<poldercast::Node> {
-        let topology = self.lock.read().await;
-        topology
-            .nodes()
-            .all_available_nodes()
-            .into_iter()
-            .cloned()
+    pub async fn list_available(&self) -> Vec<ProfileInfo> {
+        let inner = self.lock.read().await;
+        let profiles = inner.topology.peers();
+        profiles
+            .pool()
+            .iter()
+            .chain(profiles.trusted().iter())
+            .map(|(_, profile)| profile.into())
             .collect()
     }
 
-    pub async fn list_non_public(&self) -> Vec<poldercast::Node> {
-        let topology = self.lock.read().await;
-        topology
-            .nodes()
-            .all_unreachable_nodes()
-            .into_iter()
-            .cloned()
+    pub async fn list_non_public(&self) -> Vec<ProfileInfo> {
+        let inner = self.lock.read().await;
+        let profiles = inner.topology.peers();
+        profiles
+            .pool()
+            .iter()
+            .chain(profiles.trusted().iter())
+            .filter_map(|(_, profile)| {
+                if Gossip::from(profile.gossip().clone()).is_global() {
+                    None
+                } else {
+                    Some(profile.into())
+                }
+            })
             .collect()
     }
 
-    pub async fn nodes_count(&self) -> poldercast::Count {
-        let topology = self.lock.read().await;
-        topology.nodes().node_count()
+    /// register that we were able to establish an handshake with given peer
+    pub async fn promote_node(&self, node: &NodeId) {
+        let mut inner = self.lock.write().await;
+        inner.topology.promote_peer(node);
     }
 
-    /// register a strike against the given node id
-    ///
-    /// the function returns `None` if the node was not even in the
-    /// the topology (not even quarantined).
-    pub async fn report_node(&self, address: Address, issue: StrikeReason) -> Option<PolicyReport> {
-        let mut topology = self.lock.write().await;
-        topology.update_node(address, |node| {
-            node.record_mut().strike(issue);
-        })
+    /// register a strike against the given peer
+    pub async fn report_node(&self, node_id: &NodeId) {
+        let mut inner = self.lock.write().await;
+        if let Some(node) = inner.topology.get(node_id).cloned() {
+            if inner.quarantine.quarantine_node((&node).into()) {
+                inner.topology.remove_peer(node_id);
+                // Don't know what is the purpose of trusted peers in poldercast,
+                // this is a quick hack to treat those as standard ones
+                inner.topology.remove_peer(node_id);
```

Comment:
1. Please point me at the doc/cite a message about that.
2. What is the purpose of calling `remove_peer` twice?

Reply:
I'm not sure this is the intended solution, but reading the source code I understand that:
* `remove_peer` [(source)](https://github.com/primetype/poldercast/blob/c5a8603addb7e0135bbfbebce0e6ffe38c68edcf/src/topology.rs#L90) remove the peer from each of the layers and also demote it.
* peers are divided into three levels [(source)](https://github.com/primetype/poldercast/blob/c5a8603addb7e0135bbfbebce0e6ffe38c68edcf/src/profiles.rs#L6), `trusted`, `pool` and `dirty`. The only difference between these that I can see is that when receiving a gossip about a peer, if it is currently on our `dirty` pool  we do not re-add it to the topology [(source)](https://github.com/primetype/poldercast/blob/c5a8603addb7e0135bbfbebce0e6ffe38c68edcf/src/topology.rs#L116)
* You are supposed to call `promote_peer` when you can successfully connect to a peer, and that moves a peer up 1 pool (dirty -> pool -> trusted) 
* Imagine this scenario: node A connects to node B and calls `promote_peer` -> B gets promoted to the `trusted` pool. Then, for some reason, the connection drops, and we would like to quarantine B for some time. If we simply call `remove_peer` once, it gets remove from the layers, but the next gossip would re-add it right away, since it got demoted from `trusted` to `pool`. By calling `remove_peer` again, we demote it all the way to `dirty`, so that even if we receive gossips about it while it should be quarantined they will be ignored

Reply:
This really indicates a lack of proper method in poldercast. OK to use it this way, as long as the hack is documented.